idl/group0.idl.hh

```diff
@@ -24,6 +24,7 @@ enum class group0_upgrade_state : uint8_t {
     use_new_procedures = 3,
 };
 
+verb [[with_client_info, abortable]] get_group0_upgrade_state () -> service::group0_upgrade_state;
```

Comment:
What is the argument in favor of extending `peer_exchange` verb, which has **nothing** to do with upgrade - it's designed for a completely different purpose (implementing the discovery algorithm)?

We could as well extend the `group0_modify_config` verb, it makes such as much sense.

Reply:
I agree with @kostja that if possible we should avoid adding verbs. We made adding new verbs very easy, but this is a double edged sword - now their number grows fast. They are easy to add, but hard to retire and in the end you do not know which one of them are obsolete and what is still relevant.

Reply:
It is a classical train vs taxi argument. In a train you can have noisy neighbors, so there is less comfort, and one is inclined to choose comfort over everything else. But rationally one has to assign some finite weights to all pros and cons and make a balanced decision.

Reply:
> One usually doesn't have to argue for not adding a new entity to the system - this is a basic Occam's razor.

By extending an existing RPC you're adding entities as well. And since the extended part has nothing to do with the original RPC, it's only more confusing. You're only trying to sweep under the rug the fact that you just added a new entity - by merging two separate verbs into one.

> Any RPC is just a vehicle for an information request, while the request does not have to request a single bit of information per time - it is OK to request all the bits, or get a variant back as a responses. Reducing the number of round trips is also a nice benefit.

You won't reduce the number of round-trips. You would use the verb one time to do thing X, then another time to do Y, where X has nothing to do with Y. You will request for a single bit of info when doing X, then another bit of info when doing Y.

It would be like this:
```
std::variant<peer_exchange_response, get_group0_upgrade_state_response>
peer_exchange_or_get_group0_upgrade_state(std::variant<peer_exchange_args, get_group0_upgrade_state_args>)
```
or something worse.

> I agree with @kostja that if possible we should avoid adding verbs. We made adding new verbs very easy, but this is a double edged sword - now their number grows fast. They are easy to add, but hard to retire and in the end you do not know which one of them are obsolete and what is still relevant.

If there is no reason to add a verb we shouldn't add a verb. If there is a good reason, which I think there is in this case, then trying to avoid it doesn't make sense.

Whether they are hard to retire or not is IMO irrelevant, but I'll bite... how are they hard to retire? It's easy to grep the codebase if you're still calling a verb.